cgi-bin/CATS/UI/Contests.pm

```diff
@@ -97,6 +97,73 @@ sub contests_new_save {
     msg(1028, Encode::decode_utf8($c->{title}));
 }
 
+sub install_problems {
+    my ($cid, $not_installed) = @_;
+    for my $judge_id (keys $not_installed) {
+        for (@{$not_installed->{$judge_id}}) {
+            CATS::Job::create($cats::job_type_initialize_problem, {
+                judge_id => $judge_id,
+                problem_id => $_,
+                contest_id => $cid,
+            })
+        }
+    }
+    $dbh->commit;
+    # msg(???);
+}
+
+sub problems_installed_frame {
+    my ($p) = @_;
+
+    init_template($p, 'problems_installed.html.tt');
+    $p->{id} or return;
+
+    my $judges = $dbh->selectall_arrayref(q~
+        SELECT id, nick
+        FROM judges WHERE is_alive = 1~, { Slice => {} });
+
+    my $problems = $dbh->selectall_arrayref(q~
+        SELECT P.id, P.title FROM problems P
+        INNER JOIN contest_problems CP ON P.id = CP.problem_id
+        WHERE CP.contest_id = ?~, { Slice => {} },
+        $p->{id});
+
+    my $already_installed = $dbh->selectall_arrayref(qq~
+        SELECT J.id as judge_id , P.id as problem_id, JB.finish_time FROM jobs JB
+        INNER JOIN judges J on J.id = JB.judge_id
+        INNER JOIN problems P on P.id = JB.problem_id
+        INNER JOIN contest_problems CP on CP.problem_id = P.id
+        WHERE JB.state = $cats::job_st_finished AND
+            P.upload_date < JB.finish_time AND CP.contest_id = ?~, { Slice => {} },
+        $p->{id});
+
+    my $h = {};
+    $h->{$_->{judge_id}}->{$_->{problem_id}} = $_->{finish_time} for @$already_installed;
+
+    my $not_installed = {};
+    $not_installed->{$_->{judge_id}} = [] for @$judges;
+
+    my $problems_installed = [ map {
+        my $judge_id = @{$judges}[$_]->{id};
+        my $hr = $h->{$judge_id} || {};
+        {
+            judge_name => @{$judges}[$_]->{nick},
+            data => [ map {
+                push @{$not_installed->{$judge_id}}, @{$problems}[$_]->{id};
```

Comment:
1) `$problems->[$_]`
2) BUT just use normal `map`, as above
3) push ... if ?
4) Meh, this is non-functional anyway, so maybe `for`s instead of `map`s?